File: bioprocs/scripts/vcfnext/pGTMat2Plink.r
library(methods)

infile   = {{i.infile | R}}
metafile = {{i.metafile | R}}
outdir   = {{o.outdir | R}}
plink    = {{args.plink | R}}
keeptxt  = {{args.keeptxt | R}}
prefix   = file.path(outdir, {{i.infile | fn2 | R}})
pedfile  = paste0(prefix, ".ped")
mapfile  = paste0(prefix, ".map")

# column names could be:
# FID, IID, PID, MID, Sex, Pheno
if (nchar(metafile) > 0) {
	metadata    = read.table(metafile, header = T, row.names = 1, sep = "\t", check.names = F)
	metasamples = rownames(metadata)
	metacols    = colnames(metadata)
} else {
	metasamples = c()
	metacols    = c()
}

indata = read.table(infile, header = T, row.names = 1, sep = "\t", check.names = F)

samples  = colnames(indata)
snps     = rownames(indata)
nsamples = ncol(indata)
nsnps    = nrow(indata)
pedncol  = nsnps + 6
pedcols  = c('FID', 'IID', 'PID', 'MID', 'Sex', 'Pheno', snps)
comments = c('Use --compound-genotypes in later analysis')
comments = c(comments, paste(pedcols, collapse = "\t"))

mapmat = matrix(ncol = 4, nrow = nsnps)
pedmat = sapply(1:nsnps, function(i) {
	snpinfo  = unlist(strsplit(snps[i], "_"))
	mapmat[i, ] <<- c(
		if (startsWith(snpinfo[1], 'chr')) substring(snpinfo[1], 4) else snpinfo[1],
		if (startsWith(snpinfo[3], 'rs')) snpinfo[3] else paste(snpinfo[1], snpinfo[2], sep = "_"),
		0,
		snpinfo[2]
	)
	ret     = rep('00', nsamples)
	ret[which(indata[i, ] == 0)] = paste(rep(snpinfo[4], 2), collapse = '')
	ret[which(indata[i, ] == 1)] = paste(snpinfo[4:5], collapse = '')
	ret[which(indata[i, ] == 2)] = paste(rep(snpinfo[5], 2), collapse = '')
    ret
})
rm(indata)

write.table(mapmat, mapfile, sep = ' ', row.names = F, col.names = F, quote = F)
rm(mapmat)

pedmeta = matrix(NA, ncol = 6, nrow = nsamples)
rownames(pedmeta)   = samples
colnames(pedmeta)   = pedcols[1:6]
pedmeta [, 'FID']   = samples
pedmeta [, 'IID']   = samples
pedmeta [, 'PID']   = 0
pedmeta [, 'MID']   = 0
pedmeta [, 'Sex']   = 9
pedmeta [, 'Pheno'] = -9

if ('FID' %in% metacols) {
	pedmeta[metasamples, 'FID'] = metadata[metasamples, 'FID']
}
if ('PID' %in% metacols) {
	pedmeta[metasamples, 'PID'] = metadata[metasamples, 'PID']
}
if ('MID' %in% metacols) {
	pedmeta[metasamples, 'MID'] = metadata[metasamples, 'MID']
}
if ('Sex' %in% metacols) {
	pedmeta[metasamples, 'Sex'] = metadata[metasamples, 'Sex']
}
if ('Pheno' %in% metacols) {
	pedmeta[metasamples, 'Pheno'] = metadata[metasamples, 'Pheno']
}
pedmat = cbind(pedmeta, pedmat)
rm(pedmeta)

for (comment in comments) {
	cat(paste(c('#', comment, '\n')), file = pedfile, append = T)
}
write.table(pedmat, pedfile, sep = ' ', row.names = F, col.names = F, quote = F, append = T)

cmd = sprintf("%s --file %s --make-bed --out %s", plink, shQuote(prefix), shQuote(prefix))
system(cmd)

if (!keeptxt) {
	file.remove(pedfile)
	file.remove(mapfile)
}

File: bioprocs/scripts/vcfnext/pGTMat2Plink.py
from os import path
from bioprocs.utils.tsvio2 import TsvReader, TsvWriter
from bioprocs.utils import logger, shell2 as shell

infile   = {{i.infile | quote}}
metafile = {{i.metafile | quote}}
outdir   = {{o.outdir | quote}}
plink    = {{args.plink | quote}}
keeptxt  = {{args.keeptxt | repr}}
chrmaps  = {{args.chrmaps | str}}
prefix   = path.join(outdir, {{i.infile | fn2 | quote}})
tpedfile  = prefix + ".tped"
tfamfile  = prefix + ".tfam"

shell.load_config(plink = plink)

# column names could be:
# FID, IID, PID, MID, Sex, Pheno
if metafile:
	logger.info('Reading metafile ...')
	metadata = dict(TsvReader(
		metafile,
		cnames = True,
		row    = lambda r: tuple((r.IID, r))
	).dump())
else:
	metadata = None

logger.info('Reading genotype matrix ...')
# snp1 gt1s1 gt1s2 ...
inreader = TsvReader(infile, cnames = True)
samples  = inreader.meta[1:]

logger.info('Writing tfam file ...')
tfamWriter = TsvWriter(tfamfile)
tfamWriter.meta = ['FID', 'IID', 'PID', 'MID', 'Sex', 'Pheno']
#tfamWriter.writeHead(callback = lambda meta: '#' + '\t'.join(meta))
uniqueIDs = {}
if not metadata:
	for s in samples:
		tfamWriter.write([s, s, '0', '0', 'other', '-9'])
else:
	for s in samples:
		fid     = metadata[s].FID if s in metadata and 'FID' in metadata[s] else s
		idcheck = fid + ' ' + s
		if idcheck in uniqueIDs:
			raise ValueError('Duplicated ID {!r}'.format(idcheck))
		uniqueIDs[idcheck] = 1
		tfamWriter.write([
			fid,
			s,
			(metadata[s].PID or '0') if s in metadata and 'PID' in metadata[s] else '0',
			(metadata[s].MID or '0') if s in metadata and 'MID' in metadata[s] else '0',
			(metadata[s].Sex or 'other') if s in metadata and 'Sex' in metadata[s] else 'other',
			(metadata[s].Pheno or '-9') if s in metadata and 'Pheno' in metadata[s] else '-9'
		])
tfamWriter.close()

def getCompondGT(gt, ref, alt):
	compGTs = {
		"0": ref + ' ' + ref,
		"1": ref + ' ' + alt,
		"2": alt + ' ' + alt
	}
	return compGTs.get(gt, '0 0')

logger.info('Writing tped file ...')
tpedWriter = TsvWriter(tpedfile)
for r in inreader:
	(chrom, pos, _, ref, alt) = r[0].split('_')
	if chrom.startswith('chr'):
		chrom = chrom[3:]
	chrom = chrmaps.get(chrom, chrom)
	tpedWriter.write([chrom, r[0], 0, pos] + [getCompondGT(gt, ref, alt) for gt in r.values()[:]])
tpedWriter.close()

logger.info("Converting using plink ...")
shell.fg.plink({'make-bed': True}, tfile = prefix, out = prefix)

if not keeptxt:
	shell.rm_rf(tpedfile)
	shell.rm_rf(tfamfile)
